File: src/main/java/org/jdc/template/ux/individual/IndividualPresenter.java
package org.jdc.template.ux.individual;


import com.google.android.gms.analytics.HitBuilders;

import org.jdc.template.Analytics;
import org.jdc.template.model.database.main.individual.Individual;
import org.jdc.template.model.database.main.individual.IndividualManager;
import org.jdc.template.ui.mvp.BasePresenter;

import javax.inject.Inject;

public class IndividualPresenter extends BasePresenter {
    private final IndividualManager individualManager;
    private final Analytics analytics;

    private IndividualContract.View view;
    private long individualId;
    private long modelTs = 0L;

    @Inject
    public IndividualPresenter(IndividualManager individualManager, Analytics analytics) {
        this.individualManager = individualManager;
        this.analytics = analytics;
    }

    public void init(IndividualContract.View view, long individualId) {
        this.view = view;
        this.individualId = individualId;
    }

    @Override
    public void load() {
        analytics.send(new HitBuilders.EventBuilder()
                .setCategory(Analytics.CATEGORY_INDIVIDUAL)
                .setAction(Analytics.ACTION_VIEW)
                .build());
        loadIndividual();
    }

    @Override
    public void reload(boolean forceRefresh) {
        if (forceRefresh || modelTs != individualManager.getLastTableModifiedTs()) {
            loadIndividual();
        }
    }

    @Override
    public void register() {
        super.register();
    }

    @Override
    public void unregister() {
        super.unregister();
    }

    private void loadIndividual() {
        if (individualId <= 0) {
            return;
        }

        modelTs = individualManager.getLastTableModifiedTs();

        Individual individual = individualManager.findByRowId(individualId);
        if (individual != null) {
            view.showIndividual(individual);
        }
    }

    void deleteIndividualClicked() {
        view.promptDeleteIndividual();
    }

    public void deleteIndividual() {
        individualManager.delete(individualId);

        analytics.send(new HitBuilders.EventBuilder()
                .setCategory(Analytics.CATEGORY_INDIVIDUAL)
                .setAction(Analytics.ACTION_DELETE)
                .build());

        view.close();
    }

    void editIndividualClicked() {
        view.showEditIndividual(individualId);
    }
}

File: src/main/kotlin/org/jdc/template/ux/individual/IndividualPresenter.kt
package org.jdc.template.ux.individual

import com.google.android.gms.analytics.HitBuilders
import kotlinx.coroutines.experimental.Job
import kotlinx.coroutines.experimental.launch
import kotlinx.coroutines.experimental.run
import org.jdc.template.Analytics
import org.jdc.template.model.database.main.individual.IndividualManager
import org.jdc.template.ui.mvp.BasePresenter
import org.jdc.template.util.CoroutineContextProvider
import javax.inject.Inject

class IndividualPresenter @Inject
constructor(private val cc: CoroutineContextProvider,
            private val analytics: Analytics,
            private val individualManager: IndividualManager): BasePresenter() {

    private lateinit var view: IndividualContract.View
    private var individualId = 0L
    private var modelTs = 0L

    fun init(view: IndividualContract.View, individualId: Long) {
        this.view = view;
        this.individualId = individualId
    }

    override fun load(): Job? {
        analytics.send(HitBuilders.EventBuilder().setCategory(Analytics.CATEGORY_INDIVIDUAL).setAction(Analytics.ACTION_VIEW).build())
        return loadIndividual()
    }

    override fun reload(forceRefresh: Boolean): Job? {
        if (forceRefresh || modelTs != individualManager.getLastTableModifiedTs()) {
            return loadIndividual()
        }

        return null
    }

    private fun loadIndividual(): Job {
        val job = launch(cc.ui) {
            modelTs = individualManager.getLastTableModifiedTs()

            val individual = run(context + cc.commonPool) {
                individualManager.findByRowId(individualId)
            }

            if (individual != null) {
                view.showIndividual(individual)
            }
        }

        compositeJob.add(job)
        return job
    }

    fun deleteIndividualClicked() {
        view.promptDeleteIndividual()
    }

    fun deleteIndividual() {
        individualManager.delete(individualId)
        analytics.send(HitBuilders.EventBuilder().setCategory(Analytics.CATEGORY_INDIVIDUAL).setAction(Analytics.ACTION_DELETE).build())
        view.close()
    }

    fun editIndividualClicked() {
        view.showEditIndividual(individualId)
    }

}
